Eliminate extraneous copies of dht::token_range_vector

bhalevy opened the pull request:
In several places we copy token range vectors where we could move them and eliminate unnecessary memory copies.

Ref #11005

avikivity commented on dht/range_streamer.cc:
```diff
@@ -301,9 +301,6 @@ future<> range_streamer::stream_async() {
                 } catch (...) {
-                    for (auto& range : ranges_to_stream) {
-                        range_vec.push_back(range);
-                    }
```
Eventually this is part of _to_stream. How do you know it's not reused?

avikivity replied:
@asias

bhalevy replied:
> Eventually this is part of _to_stream. How do you know it's not reused?

I inspected all call sites

asias replied:
We will report wrong nr_ranges_remaining. After this patch, the failed ranges will be reported as ranges finished successfully.

asias replied:
The commit log says:

```
    dht/range_streamer: stream_async no need to return ranges to range_vec on exception
    
    There is no retry when stream_async fails
    so no need to return the failed ranges to range_vec.
    
    This will make it easier to move ranges_to_stream
    does the stack rather than copying it in the following
    patches.
```
 
However, I think range_streamer::stream_async() should keep the failed ranges instead of dropping it silently, we do not retry in stream_async (any more) does not mean the caller of the stream_async() woulc not do the retry it self.

avikivity commented:
@asias please review, it's very tricky code

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3534/

bhalevy commented:
ping @asias, please review

bhalevy commented:
@asias ping review

asias commented on dht/range_streamer.cc:
```diff
@@ -288,22 +286,24 @@ future<> range_streamer::stream_async() {
                             continue;
                         } else {
-                            do_streaming();
+                            do_streaming(std::move(ranges_to_stream));
```
The ranges_to_stream is moved here, but we are accessing it again at line 294 in the loop.

bhalevy replied:
Right. The content of the vector is moved to `do_streaming` and it becomes empty.
The access at line 294 when we loop checks the vector after inserting more ranges to it at line 292 so I see no problem with that.

asias replied:
IIRC, a moved from vector is valid but unspecified. It is not guaranteed to be empty. 

E.g., see https://stackoverflow.com/questions/17730689/is-a-moved-from-vector-always-empty

asias replied:
@avikivity How do you think? 

How about this?
```
do_streaming(std::move(ranges_to_stream));
ranges_to_stream = {};
```

bhalevy replied:
you're right.
https://en.cppreference.com/w/cpp/language/move_constructor even explicitly mentions thr std::vector case.

bhalevy replied:
I think that `.clear()` would be clearer than `= {}` :)

bhalevy replied:
fixed in v3

asias replied:
I am not sure why .clear is better.  Calling .clear after the vector is moved is still accessing  a removed from object (which we should avoid in general). The removed from vector is still in a unspecified state (Yes,  clear makes .size() return zero and invalidate references and so do). 

In stead

ranges_to_stream = {};

makes it clear, we assign a new object to ranges_to_stream, we do not care about the old one anymore (which is removed from vector).

@avikivity ?

bhalevy replied:
@asias there should be no provlem accessing the moved-from vector. The standard says it ends up in an unspecified, but valid, state.  So it's not guranteed to be empty, but it may be accessed for e.g. clear().  If it would leave the capacity of the moved-frim vector by moving all the objects from it, then clear will save an allocation. Assiging an empty vector will guarantee allocation (though in practice the moved vector does lose its capacity and allocation will happen if we insert more tokens in the loop)

asias replied:
IMHO, it is better to follow the simple rule: never touch a moved from object than playing games to save an allocation.
The code is on the control plane anyway.

bhalevy replied:
This rule is an overkill.  Move semantics require that rhe moved-from object will remain valid, otherwise destroying it will easily end up with undefined behavior since the destructor is always called on it before it's destroyed.
I agree that if its state is unspecified the code shouldn't make any assumptions about its state post-move, but I see no provlem in getting it back into a specified state.

Consider a moved-from optional, which is used quite extensively in our code. You know it still evaluates to true as it is still engaged, holding a moved-from object, any in many cases, we reset this moved-from optional to get it into the disengaged state.

bhalevy replied:
Also, @asias, assigning the moved-from object with a default-constructed object is touching it, as it executes the move-assignment operator which typically destroys the assigned object before move-constructing it from the rvalue object.

asias replied:
Benny, it's a good point that  default constructor is touching it too. It is fair enough. It's up to the maintainers to decide.

avikivity replied:
My preference is to use std::exchange() when you want to ensure the object is in a known state after moving it.

bhalevy replied:
I don't think that works for exception safety as std::exchange takes effect before the rvalue is passed to the called function, and if throws, it's lost.

I'm going to send another version though with improved exception safety that should be less contentious around clearing the ranges vector past move

asias commented:
Sorry, I missed the ping. I added some comments.

bhalevy commented:
In v2 ([ef6c38a](https://github.com/scylladb/scylladb/commit/ef6c38a91c4bd3a513cf8266d749b8e1542d37c7)):
- dropped:
  - dht/range_streamer: stream_async no need to return ranges to range_vec on exception
  - dht/range_streamer: stream_async: no need to erase from range_vec one entry at a time
- Instead, added:
  - dht/range_streamer: stream_async: erase from range_vec only after do_streaming success

asias commented on dht/range_streamer.cc:
```diff
@@ -288,22 +286,26 @@ future<> range_streamer::stream_async() {
                 };
                 try {
+                  if (range_vec.size() <= nr_ranges_per_stream_plan) {
+                    do_streaming(std::move(range_vec));
```
Same here,  we might range_vec after it is moved.

bhalevy replied:
fixed in v3

bhalevy commented:
In v3 ([0cac873](https://github.com/scylladb/scylladb/commit/0cac8739b2add30c03218b4b9c3f4e8b49c8217f)):
- dht/range_streamer: stream_async: move ranges_to_stream to do_streaming
  - clear `ranges_to_stream` after std::move in loop
- dht/range_streamer: stream_async: erase from range_vec only after do_streaming success
  - `it = range_vec.erase` since it is invalidated by the erase
  - `range_vec.clear()` after the eventual call to do_streaming, since items are not erased one at a time with this patch
- dht/range_streamer: stream_async: optimize small range_vec
  - call `range_vec.clear()` on success

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3771/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3772/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3775/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3773/

bhalevy commented:
In v4 ([6b9a2c7](https://github.com/scylladb/scylladb/commit/6b9a2c755fd92dd86172008745def347bf626da7)):
- rebased
- cleaned up exception safety dependencies
- reordered erasing entries from `range_vec` patch before moving `ranges_to_stream` to `do_streaming`
  - this allows using `std::exchange(ranges_to_stream, {})` for calling `do_streaming`.
  - but it still can't be used for `do_streaming(std::move(range_vec))` for exception safety reasons
    - comment added about that.

bhalevy commented:
@avikivity ping review

avikivity commented:
@asias are you done with your review?

bhalevy commented:
ping @asias

asias commented on dht/range_streamer.cc:
```diff
@@ -274,36 +273,45 @@ future<> range_streamer::stream_async() {
+                    // Do not exchange range_vec with {} for exception safety reasons.
+                    // ip_range_vec.second is used in nr_ranges_to_stream
+                    // to calculate nr_ranges_remaining, so it should be cleared
+                    // only on success.
```
do_streaming(std::move(range_vec));
range_vec.clear();

Here, we move range_vec before it has finished successfully.  ragne_vec.size will return zero even if do_streaming has failed.

bhalevy replied:
Not any more, since do_streaming accepts and passes its `ranges_to_stream` parameter as an rvalue reference so the move takes place eventually only in `request_ranges -> add_stream_request` or `transfer_ranges -> add_transfer_ranges` and those were made exception safe in this series.

bhalevy replied:
okay, removed the 'optimize small range_vec' patch from v5

asias commented:
Sorry for the delay. I was out on vacation. I had one comment on the do_streaming(std::move(range_vec)).

bhalevy commented:
v5 ([083e8bf](https://github.com/scylladb/scylladb/commit/083e8bf2b828ae4d6e731cbfd5b36c59e69d9334)):
- rebased
- removed "dht/range_streamer: stream_async: optimize small range_vec"

asias commented:
@avikivity Looks good to me. Please consider merging it.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4056/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4057/

bhalevy commented:
[7961e8e](https://github.com/scylladb/scylladb/commit/7961e8e1208147d4ba0daccd2d7091360e020dcf): rebased

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4290/

bhalevy commented:
ci keeps failing in test_cluster_reduction_with_cdc[Multi_DC_cluster] but I couldn't reproduce it locally

bhalevy commented:
In v6 ([4cbaa69](https://github.com/scylladb/scylladb/commit/4cbaa69b8c483662f8e38e214e243cbb8b2c9334)):
- rebased
- fixed some lvalue/rvalue move-related nits

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4550/

bhalevy commented:
@scylladb/dtest-maintainers please consider merging

denesb commented:
I'm not a dtest maintainer but I'll go ahead and queue this as it appears all have approved.

bhalevy commented:
> I'm not a dtest maintainer

Oops, sorry.  I meant `scylla-maintainers` of course :)

kbr-scylla commented:
Dequeued. `test_cluster_reduction_with_cdc` and `test_cluster_expansion_with_cdc` are currently failing on `next`. One of the failures is similar to https://github.com/scylladb/scylladb/issues/11531, which happened because a bug was introduced in replica set calculation.

It appears that this is the PR where the failures reappeared, and it modifies token-ring related code, so it looks like the culprit.

kbr-scylla commented on dht/range_streamer.cc:
```diff
@@ -214,15 +214,15 @@ void range_streamer::add_rx_ranges(const sstring& keyspace_name, std::unordered_
 
 // TODO: This is the legacy range_streamer interface, it is add_rx_ranges which adds rx ranges.
-future<> range_streamer::add_ranges(const sstring& keyspace_name, locator::effective_replication_map_ptr erm, dht::token_range_vector ranges, gms::gossiper& gossiper, bool is_replacing) {
+future<> range_streamer::add_ranges(const sstring& keyspace_name, locator::effective_replication_map_ptr erm, dht::token_range_vector&& ranges, gms::gossiper& gossiper, bool is_replacing) {
```
Adding `&&` here doesn't eliminate a copy constructor invocation, but it does eliminate a move constructor invocation (when calling `add_ranges`), which in terms of performance shouldn't really make a difference. It does make the code a bit harder to reason about though, because now the parameter refers to a variable living outside the function. Although it's immediately moved into the lambda so it's still pretty simple.

bhalevy replied:
It makes the function exception safe, e.g. if allocating the function that captures `ranges` throws bad_alloc.

bhalevy replied:
gone in v7

kbr-scylla commented on streaming/stream_request.hh:
```diff
@@ -29,12 +29,12 @@ public:
     std::vector<sstring> column_families;
     stream_request() = default;
-    stream_request(sstring _keyspace, dht::token_range_vector _ranges, std::vector<sstring> _column_families)
+    stream_request(sstring _keyspace, dht::token_range_vector&& _ranges, std::vector<sstring>&& _column_families) noexcept
```
Re commit message: `std::vector` move constructor is `noexcept` since C++17 so the `stream_request` ctor could already have been made `noexcept` earlier.

kbr-scylla replied:
And you're invoking `std::vector` move constructor below anyway (`ranges(std::move(_ranges))` etc.), so you are depending on the `noexcept`ness of `std::vector`'s move constructor anyway if you want to mark this ctor `noexcept`.

bhalevy replied:
true

kbr-scylla commented on streaming/stream_session.hh:
```diff
@@ -215,7 +215,7 @@ public:
      * @param columnFamilies ColumnFamily names. Can be empty if requesting all CF under the keyspace.
      */
-    void add_stream_request(sstring keyspace, dht::token_range_vector ranges, std::vector<sstring> column_families) {
+    void add_stream_request(sstring keyspace, dht::token_range_vector&& ranges, std::vector<sstring>&& column_families) {
```
I don't think this affects exception safety in any way.

bhalevy replied:
If `_requests.emplace_back` fails to allocate, the `ranges` arg, that was already moved by the caller, will be destroyed, while with passing the rvalue reference the move will not take place in this case.
If allocation succeeded, then `std::move(ranges)` will take place in the ctor (hence emplace_back is preferred over push_back) which is noexcept at this point.

kbr-scylla commented on dht/range_streamer.cc:
```diff
@@ -274,7 +274,7 @@ future<> range_streamer::stream_async() {
                     nr_ranges_streamed += ranges_to_stream.size();
                     if (_nr_rx_added) {
-                        sp.request_ranges(source, keyspace, ranges_to_stream);
+                        sp.request_ranges(source, keyspace, std::move(ranges_to_stream));
```
I think this is actually making the code exception unsafe (at least more unsafe than it was):
if `sp.execute().discard_result().get()` below throws, we will enter the `catch (...) { ... }` block later below, which will iterate over `ranges_to_stream` - which has now been (incorrectly) moved from.

bhalevy replied:
`std::move` does not call anything at this point, it just passes an rvalue reference down the stack.
The point in exception safety is that if an exception is thrown, the move will NOT take place.

bhalevy replied:
that's a good point.
maybe the caller needs to keep the ranges alive throughout the stream plan instead of moving them and pass a reference to the vector the the stream_plan instead, and finally consume them iff the stream_plan succeeded.

bhalevy replied:
in v7 this token range vector contains a discardable copy.
the changes to `_to_stream` are safe as they are applied only on success.

kbr-scylla commented on streaming/stream_session.cc:
```diff
@@ -490,7 +490,7 @@ void stream_session::add_transfer_ranges(sstring keyspace, dht::token_range_vect
         auto it = _transfers.find(cf_id);
         if (it == _transfers.end()) {
-            stream_transfer_task task(shared_from_this(), cf_id, ranges);
+            stream_transfer_task task(shared_from_this(), cf_id, std::move(ranges));
```
I think `ranges` might be used again in a later iteration of this `for (auto& cf : cfs)` loop, in which case this would be use-after-move.

raphaelsc replied:
True. If we really want to avoid copy of ranges here (multiple copies can cause stall on large cluster installations right?), perhaps a `lw_shared_ptr<const dht::token_range_vector>` which is shared by all instances of stream_transfer_task?!

bhalevy replied:
I doubt that the lw_shared_ptr is required.
We don't need to leave it behind for background processing.
`range_streamer::stream_async` starts a thread so it keep easily hold on to the ranges while the streaming_plan executes, and it waits for it anyhow.

kbr-scylla commented on streaming/stream_session.cc:
```diff
@@ -494,7 +494,7 @@ void stream_session::add_transfer_ranges(sstring keyspace, dht::token_range_vect
             assert(inserted);
         } else {
-            it->second.append_ranges(ranges);
+            it->second.append_ranges(std::move(ranges));
```
I think `ranges` might be used again in a later iteration of this `for (auto& cf : cfs)` loop, in which case this would be use-after-move.

kbr-scylla commented on streaming/stream_transfer_task.cc:
```diff
@@ -239,8 +239,11 @@ future<> stream_transfer_task::execute() {
 }
 
-void stream_transfer_task::append_ranges(const dht::token_range_vector& ranges) {
-    _ranges.insert(_ranges.end(), ranges.begin(), ranges.end());
```
`std::vector::insert` already gives a strong exception guarantee, I think it's harder to see that the new code also gives it (better to rely on stdlib guarantees than enforce it on our own).

https://en.cppreference.com/w/cpp/container/vector/insert

bhalevy replied:
I think that's doable

kbr-scylla commented on streaming/stream_plan.cc:
```diff
@@ -35,7 +35,7 @@ stream_plan& stream_plan::transfer_ranges(inet_address to, sstring keyspace, dht
     _range_added = true;
     auto session = _coordinator->get_or_create_session(_mgr, to);
-    session->add_transfer_ranges(keyspace, std::move(ranges), std::move(column_families));
+    session->add_transfer_ranges(keyspace, std::move(ranges), column_families);
```
I don't get this change. What was wrong with moving `column_families`? It's a local parameter in `transfer_ranges`, we don't use it later, so we can `move` it.

kbr-scylla commented on dht/range_streamer.cc:
```diff
@@ -276,7 +276,7 @@ future<> range_streamer::stream_async() {
                         sp.request_ranges(source, keyspace, std::move(ranges_to_stream));
                     } else if (_nr_tx_added) {
-                        sp.transfer_ranges(source, keyspace, ranges_to_stream);
+                        sp.transfer_ranges(source, keyspace, std::move(ranges_to_stream));
```
As above (and why are the two changes in separate commits - they seem very similar)

bhalevy replied:
squashed together in a single commit in v7

kbr-scylla commented on dht/range_streamer.cc:
```diff
@@ -290,20 +290,19 @@ future<> range_streamer::stream_async() {
                 try {
                     for (auto it = range_vec.begin(); it < range_vec.end();) {
                         ranges_to_stream.push_back(*it);
-                        it = range_vec.erase(it);
```
Ah, so in an earlier commit you introduce the use-after-move in the calls to `request_ranges` and `transfer_ranges` above, and here you fix it.
Instead, this commit should go first.

kbr-scylla commented on dht/range_streamer.cc:
```diff
@@ -290,20 +290,19 @@ future<> range_streamer::stream_async() {
                             continue;
                         } else {
                             do_streaming();
+                            it = range_vec.erase(range_vec.begin(), it);
```
`do_streaming()` is calling `nr_ranges_to_stream()`, which is looking at these `range_vec`s, so it assumes that the `range_vec`s are modified before `do_streaming()` is called.

bhalevy replied:
Ouch.  Maybe we need to rely on the local `nr_ranges_remaining` variable instead.
Will look into that...

bhalevy replied:
added a `_nr_ranges_remaining` member that is incrementally updated.

kbr-scylla replied:
Could you include that member in this first commit?

kbr-scylla commented:
> ci keeps failing in test_cluster_reduction_with_cdc[Multi_DC_cluster]

This should have been investigated before queuing the PR. The test is not known to be flaky. If there was a suspicion that it is flaky and the failure is not caused by the PR, an issue should have been opened.

bhalevy commented:
> > ci keeps failing in test_cluster_reduction_with_cdc[Multi_DC_cluster]
> 
> This should have been investigated before queuing the PR. The test is not known to be flaky. If there was a suspicion that it is flaky and the failure is not caused by the PR, an issue should have been opened.

Of course.  I wasn't able to reproduce the dtest failure locally, and after https://github.com/scylladb/scylladb/pull/12344#issuecomment-1439595393 CI passed
so I had the false impression that the issue was dealt with.

bhalevy commented:
In v7 ([06a0902](https://github.com/scylladb/scylladb/commit/06a0902708ca2df3fb3d99a309809a91ce046764)):
- fixed use-after move in previous version in add_transfer_ranges
  - the ranges were potentially moved to multiple tables, but they should be copied, not moved.
- rebased and cleaned up series
- reordered patches so that the following come first:
  - dht/range_streamer: stream_async: erase from range_vec only after do_streaming success
  - dht/range_streamer: stream_async: incrementally update _nr_ranges_remaining
- reduced code churn by eliminating changes related to exception safety
  - as this is not strictly needed with the first patches
  - keeping token_range_vector in function signatures (instead of rvalue references)
  - no need for zealous noexept specifiers
- Added:
  - dht/range_streamer: get_range_fetch_map: reduce copies
  - dht/range_streamer: add_ranges: clear_gently ranges_for_keyspace
  - streaming: stream_session: maybe_yield
- dht/range_streamer: stream_async: move ranges_to_stream to do_streaming
  - revised commit message and motivation
  - it is about readability now rather than correctness / exception safety

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4661/

bhalevy commented:
@kbr-scylla please re-review and merge if fine

kbr-scylla reviewed (commented):
I'll review tomorrow morning

kbr-scylla commented on dht/range_streamer.cc:
```diff
@@ -266,8 +266,7 @@ future<> range_streamer::stream_async() {
                 size_t nr_ranges_per_stream_plan = nr_ranges_total / 10;
-                dht::token_range_vector ranges_to_stream;
-                auto do_streaming = [&] {
+                auto do_streaming = [&] (dht::token_range_vector&& ranges_to_stream) {
```
Would be even simpler to reason about if you used a value parameter instead of reference. The code wouldn't have to rely on temporary object lifetime extension.

bhalevy replied:
Fair point.
I see you already queued the PR.
If needed, I'll fix it as a follow up

kbr-scylla replied:
I think this is fine as well.

tangcheng999 commented:
@bhalevy I'm happy to see that this fix has been applied in version 5.4 and merged in master, but I didn't see the fix in version 5.2, so I want to know if I need to backport this fix to 5.2.

bhalevy commented:
> @bhalevy I'm happy to see that this fix has been applied in version 5.4 and merged in master, but I didn't see the fix in version 5.2, so I want to know if I need to backport this fix to 5.2.

@tangcheng999, yes, it was decided against backporting it to 5.2, along with other patches.
See https://github.com/scylladb/scylladb/issues/11005#issuecomment-1781176176

If you're backporting this patch to your own branch, just make sure there are no further fixes on top of it that might be required.  I don't remember off the top if my head if we had any, or if everything was dealt with in the review process.
